### PR TITLE
DRILL-7717: Support Mongo extended types in V2 JSON loader

### PR DESCRIPTION
**Pull request by paul-rogers**
# [DRILL-7717](https://issues.apache.org/jira/browse/DRILL-7717): Support Mongo extended types in V2 JSON loader

## Description

As it turns out, Drill's existing "V1" JSON reader provides support for [V1 of Mongo's extended JSON types](https://docs.mongodb.com/manual/reference/mongodb-extended-json-v1/). This PR adds that support, along with the updated [V2](https://docs.mongodb.com/manual/reference/mongodb-extended-json/) support, to the V2 JSON reader. Drill provides a number of Drill extensions to the Mongo extended types; those are also include in the V2 reader.

The V2 reader design is based on making the per-record read path as efficient as possible. Generally, we try to be two "virtual functions" away from converting a JSON token to a value vector entry. The first function is to a "value listener" that knows how to convert a JSON token to a desired data type, and a second function to write that value into a vector using a column writer.

In this design, most of the complexity (of which there is an abundant amount) is pushed into "factory" classes and method which build up a parser/listener pair appropriate for each bit of JSON structure. The prior structure was a bit rigid and assumed a direct mapping from JSON structure to Drill's data types (a JSON object structure always represented a Drill `MAP` and so on.) This PR revises that structure to be more flexible in order to handle the Mongo extended types which treat a JSON object as a holder for a type declaration and a value. You'll see lots of code changed, but much was just moving things around.

A new package is added to support the Mongo types (and Drill's extensions to the Mongo extended types.) Since this support required supporting a wider array of data types in JSON, the code also enables the use of a provided schema to do the same conversion with "non-extended" JSON.

The implementation should allow adding other forms of extended JSON if there is a need.

## Documentation

Drill's [JSON documentation](http://drill.apache.org/docs/json-data-model/) does not discuss the extended types. A full description, suitable for converting to user documentation, can be found in a [package-info.json](https://github.com/paul-rogers/drill/blob/f32057234935c632e740173d197b5104a08fa46d/exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/extended/package-info.java) file.

Once this code replaces the JSON reader (see below), users can provide a schema to specify the column type:

* `INT`, BIGINT` -- Conversion from JSON integers and strings.
* `FLOAT4`, `FLOAT8` -- Conversion from JSON integers, floats and strings. `NaN`, `-Infinity` and `Infinity` are supported as literals if enables. They are always supported as strings.
* `VARDECIMAL` -- Conversion from JSON strings, integers and floats. Conversion is done from the value string to avoid rounding errors and range limits.
* `VARCHAR` -- Conversion is allowed from any scalar type. The literal JSON text value is used.
* `VARBINARY` -- Conversion from a JSON string encoded using Bse64 as per the Mongo spec.
* `DATE` -- Allows a date, expressed as a string, in the ISO `YYYY-MM-DD` format.
* `TIME` -- Allows a time, expressed as a string, in the ISO `HH:MM:SS.SSS` format.
* `TIMESTAMP` -- Allows a date/time, expressed as an ISO string in the `YYYY-MM-DDTHH:MM:SS.SSS` format. The time is converted from the timezone expressed in the string into the local timezone (as required for a `TIMETAMP`.)
* `INTERVAL` -- String encoded in the ISO period format: `PxYxMxDTxHxMxS`.
* `INTERVALYEAR` -- String encoded in the ISO period format: `PxYxM-`.
* `INTERVALDAY` -- String encoded in the ISO period format: `PxYxxDTxHxMxS`.

## Testing

Added a category for JSON-related tests. Added unit tests for the new functionality.

The prior version had separate tests for the JSON "parser" and "loader". After these changes, maintaining that separation became harder, and less useful. The prior "parser" tests are removed with functionality moving into the "loader" tests.

No code other than the HTTP storage plugin uses this code at present, so this change will not affect other unit or functional tests.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/extended/ExtendedTypeFieldFactory.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+  private BaseExtendedValueParser numberLongParser(FieldDefn fieldDefn, boolean isArray) {
+    return new SimpleExtendedValueParser(
+        fieldDefn.parser(), ExtendedTypeNames.LONG,
+        new StrictBigIntValueListener(loader(),
+            fieldDefn.scalarWriterFor(MinorType.BIGINT, isArray)));
+  }
+
+  private BaseExtendedValueParser numberDecimalParser(FieldDefn fieldDefn, boolean isArray) {
+    // No information about precision and scale, so guess (38, 10).
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/extended/ExtendedTypeFieldFactory.java:146]: Is it possible to obtain information about scale and precision from `fieldDefn`? For the case of the schema provisioning users should be able to specify target scale and precision.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/extended/MongoBinaryValueParser.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      JsonToken token = tokenizer.requireNext();
+      if (token == JsonToken.END_OBJECT) {
+        break;
+      } else if (token != JsonToken.FIELD_NAME) {
+        throw syntaxError();
+      }
+      switch (tokenizer.textValue()) {
+        case "base64":
+          if (sawData) {
+            syntaxError();
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/extended/MongoBinaryValueParser.java:130]: Looks like `throw` is missed here and in other places.
```suggestion
            throw syntaxError();
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/loader/EmptyArrayFieldParser.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+   * column, writer and parser to replace this parser.
+   */
+  @Override
+  protected ElementParser resolve(TokenIterator tokenizer) {
+    tupleParser.loader().removeNullMarker(this);
+    if (tokenizer.peek() == JsonToken.START_ARRAY) {
+
+      // The value is [], [ foo, so resolve directly
+      return tupleParser.resolveField(key, tokenizer);
+    } else {
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/loader/EmptyArrayFieldParser.java:59]: Could you please remove the empty block?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/loader/JsonLoaderImpl.java [hunk 4]
```diff
@@ -222,18 +232,41 @@ private JsonLoaderImpl(JsonLoaderBuilder builder) {
...
             .fromStream(builder.stream)
             .fromReader(builder.reader)
             .options(builder.options)
-            .rootListener(rowListener)
+            .parserFactory(new ParserFactory() {
+              @Override
+              public ObjectParser rootParser(JsonStructureParser parser) {
+                return new TupleParser(parser, JsonLoaderImpl.this, rsLoader.writer(), builder.providedSchema);
+              }
+            })
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/loader/JsonLoaderImpl.java:249]: Please replace it with lambda:
```suggestion
            .parserFactory(parser -> new TupleParser(parser, JsonLoaderImpl.this, rsLoader.writer(), builder.providedSchema))
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/FieldParserFactory.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+ * this class reports that it is an object, then the listener should
+ * raise an exception.
+ * <p>
+ * Also, the parser cannot enforce type consistency. This method
+ * looks only at the first appearance of a value: a sample size of
+ * one. JSON allows anything.
+ * The listener must enforce semantic rules that say whether a different
+ * type is allowed for later values.
+ *
+ * @param key the name of the field
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/FieldParserFactory.java:61]: `@param` tags shouldn't appear in class javadocs...

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/ObjectParser.java [hunk 6]
```diff
@@ -82,26 +100,83 @@
...
+  /**
+   * Called at the start of a set of values for an object. That is, called
+   * when the structure parser accepts the <code>{</code> token.
+   */
+  protected void onStart() { }
+
+  /**
+   * The structure parser has just encountered a new field for this
+   * object. This method returns a parser for the field, along with
+   * an optional listener to handle events within the field. THe field typically
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/ObjectParser.java:126]:
```suggestion
   * an optional listener to handle events within the field. The field typically
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/values/UtcTimestampValueListener.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ * <p>
+ * In Shell mode, {@code <date>} is the JSON representation of a 64-bit signed
+ * integer giving the number of milliseconds since epoch UTC.
+ * </quote>
+ * <p>
+ * Drill dates are in the local time zone, so conversion is needed.
+ */
+public class UtcTimestampValueListener extends ScalarListener {
+
+  private final ZoneId localZoneId = ZoneId.systemDefault();
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/values/UtcTimestampValueListener.java:44]: It is safe to make it static.

**Comment by paul-rogers** [comment]
Added provided schema support for extended types, along with tests.

Rebased on the latest master and resolved conflicts from parallel JSON work.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/values/UtcTimestampValueListener.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ * <p>
+ * In Shell mode, {@code <date>} is the JSON representation of a 64-bit signed
+ * integer giving the number of milliseconds since epoch UTC.
+ * </quote>
+ * <p>
+ * Drill dates are in the local time zone, so conversion is needed.
+ */
+public class UtcTimestampValueListener extends ScalarListener {
+
+  private static final ZoneId localZoneId = ZoneId.systemDefault();
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/values/UtcTimestampValueListener.java:44]: Please rename it to upper case.

**Comment by paul-rogers** [comment]
@vvysotskyi, thanks much for the review. Addressed your remaining comments.

**Comment by paul-rogers** [comment]
Rebased on latest master. Squashed commits.

**Pull request merged**